### PR TITLE
[installinator] Add checks on sha256 of artifacts

### PR DESCRIPTION
**Pull request by jgallagher**
This PR adds two checksum checks:

1. After downloading artifacts from wicketd, confirm the hash of the data we received matches the hash we requested. (This applies to both host OS and control plane artifacts.)
2. After writing the host OS to disk, reread the file and hash it, ensuring it matches the host phase 2 hash we downloaded (and checked in step 1).

We could do a similar check on each control plane file, but we'd have to hash them all individually to know what to expect.

If/when we see #3441 again, this should provide an extra data point that at least when installinator was alive, it believed the host OS had been successfully written to disk.

On `madrid`, this step takes adds about 11 seconds (per slot) to the time to mupdate:

validate-host

### CODE REVIEW

File: installinator/src/dispatch.rs [hunk 1]
```diff
@@ -203,6 +204,13 @@ impl InstallOpts {
                         fetch_artifact(&cx, &host_phase_2_id, discovery, log)
                             .await?;
 
+                    check_downloaded_artifact_hash(
+                        "host phase 2",
+                        host_phase_2_artifact.artifact.clone(),
+                        host_phase_2_id.hash,
+                    )
+                    .await?;
```
Comment (sunshowers) [inline installinator/src/dispatch.rs:212]: Do we want to retry the operation if it fails? (If the answer is "not for now, maybe in the future", could you add a comment about it here?)

**Review by sunshowers — APPROVED** [review]

File: installinator/src/dispatch.rs (hunk 1 above)
Comment (jgallagher) [reply installinator/src/dispatch.rs:212]: I think not, maybe ever? We're fetching this data from wicketd (in memory) over TCP to installinator (in memory), so I think the by-far-most-likely reason for this to fail is if wicketd sends the wrong artifact, which I wouldn't expect to be resolved by a retry (and we would definitely want to know immediately if it happened). I'll add a comment though :+1:.

Comment (jgallagher) [reply installinator/src/dispatch.rs:222]: Added explanatory comments in [8ea3902](https://github.com/oxidecomputer/omicron/pull/3455/commits/8ea3902187af7918fd2e469becca7e7dcb9d604f)

File: installinator/src/write.rs [hunk 2]
```diff
@@ -385,24 +387,101 @@ impl<'a> SlotWriteContext<'a> {
...
+        })
+        .await
+        .unwrap()
+        .map_err(WriteError::ChecksumValidationError)?;
+
+        if computed_hash == self.artifacts.host_phase_2_id.hash {
+            StepSuccess::new(())
+                .with_message(format!(
+                    "validated hash {computed_hash} \
+                     for host phase 2 written to {slot:?}"
```
Comment (leftwo) [inline installinator/src/write.rs:475]: Thank you for this!

**Pull request merged**